apps/rush-lib/src/cli/logic/InstallManager.ts

```diff
@@ -526,201 +590,197 @@ export class InstallManager {
 
-      // If "--clean" or "--full-clean" was specified, or if the last install was interrupted,
-      // then we will need to delete the node_modules folder.  Otherwise, we can do an incremental
-      // install.
-      const deletingNodeModules: boolean = installType !== InstallType.Normal || !markerFileExistedAndWasValidAtStart;
+    // Based on timestamps, can we skip this install entirely?
+    if (shrinkwrapIsUpToDate && !deletingNodeModules) {
+      const potentiallyChangedFiles: string[] = [];
 
-      // Based on timestamps, can we skip this install entirely?
-      if (!deletingNodeModules) {
-        const potentiallyChangedFiles: string[] = [];
+      // Consider the timestamp on the node_modules folder; if someone tampered with it
+      // or deleted it entirely, then we can't skip this install
+      potentiallyChangedFiles.push(commonNodeModulesFolder);
 
-        // Consider the timestamp on the node_modules folder; if someone tampered with it
-        // or deleted it entirely, then we can't skip this install
-        potentiallyChangedFiles.push(commonNodeModulesFolder);
+      // Additionally, if they pulled an updated npm-shrinkwrap.json file from Git,
+      // then we can't skip this install
+      potentiallyChangedFiles.push(this._rushConfiguration.committedShrinkwrapFilename);
 
-        // Additionally, if they pulled an updated npm-shrinkwrap.json file from Git,
-        // then we can't skip this install
-        potentiallyChangedFiles.push(this._rushConfiguration.committedShrinkwrapFilename);
+      // Also consider timestamps for all the temp tarballs. (createTempModulesAndCheckShrinkwrap() will
+      // carefully preserve these timestamps unless something has changed.)
+      // Example: "C:\MyRepo\common\temp\projects\my-project-2.tgz"
+      potentiallyChangedFiles.push(...this._rushConfiguration.projects.map(x => {
+        return this._getTarballFilePath(x);
+      }));
+
+      // NOTE: If commonNodeModulesMarkerFilename (or any of the potentiallyChangedFiles) does not
+      // exist, then isFileTimestampCurrent() returns false.
+      if (Utilities.isFileTimestampCurrent(this._commonNodeModulesMarker.path, potentiallyChangedFiles)) {
+        // Nothing to do, because everything is up to date according to time stamps
+        return;
+      }
+    }
 
-        // Also consider timestamps for all the temp tarballs. (createTempModulesAndCheckShrinkwrap() will
-        // carefully preserve these timestamps unless something has changed.)
-        // Example: "C:\MyRepo\common\temp\projects\my-project-2.tgz"
-        potentiallyChangedFiles.push(...this._rushConfiguration.projects.map(x => {
-          return this._getTarballFilePath(x);
-        }));
+    // Since we're going to be tampering with common/node_modules, delete the "rush link" flag file if it exists;
+    // this ensures that a full "rush link" is required next time
+    Utilities.deleteFile(this._rushConfiguration.rushLinkJsonFilename);
 
```

Comment:
Shouldn't this just be the flag thing?

Reply:
It should be, but today it actually contains a full data structure used by "rush rebuild" to determine the build order. Nick and I recently realized that this whole data structure can be eliminated, but that's outside the scope of this PR.

---
In reply to: [187459499](https://github.com/Microsoft/web-build-tools/pull/646#discussion_r187459499) [](ancestors = 187459499)